evm/src/cpu/kernel/assembler.rs

```diff
@@ -0,0 +1,149 @@
+use std::collections::HashMap;
+
+use super::ast::PushTarget;
+use crate::cpu::kernel::{
+    ast::{Function, Item},
+    opcodes::{opcode_ordinal, push_ordinal},
+};
+
+/// The number of bytes to push when pushing an offset within the code (i.e. when assembling jumps).
+/// Ideally we would automatically use the minimal number of bytes required, but that would be
+/// nontrivial given the circular dependency between an offset and its size.
+const BYTES_PER_OFFSET: usize = 3;
+
+#[derive(PartialEq, Eq, Debug)]
+#[allow(dead_code)] // TODO: Should be used once witness generation is done.
+pub struct Kernel {
+    code: Vec<u8>,
```

Comment:
In hindsight, based on @nbgl's suggestion of adding an extra bit to distinguish privileged opcodes, it seems like these instructions could be 9 bits each so we should probably use `Vec<u16>`. Does that sound right?

Reply:
You _could_ do it that way (actually expanding the opcodes to 9 bits) but that’s not quite what I meant.

I’m imagining a bit that’s implicit in the CPU. It’s 1 if we’re in a kernel function and 0 otherwise. Like the protection ring in a CPU.

Each of the two protection rings has its own set of 8-bit opcodes (although likely with a lot of overlap, for convenience and efficiency). The “9th bit” is just the internal flag that switches between those two sets.

So the assembly for kernel code would still produce 8-bit opcodes, they would just potentially be different from userspace assembly.